rfcs/0007-handling-corruption.md

```diff
@@ -217,108 +209,184 @@ durability and availability systems on top of SlateDB.
-and SlateDB detects corruption while searching for the key, an error should be
-reported. However, if SlateDB encounters corruption while searching for AND
-finding a key, it should not return an error. This ensures the database remains
-available even in the presence of corruption.
-
-The error returned should be a distinct type which can easily be matched to
-determine if the error is due to non transient corruption or a transient error.
-(network connectivity, etc..)
-
-The rationale behind this approach is rooted in the assumption that SlateDB's
-design is a manifestation of a physical system. Linux system administrators
-would not use physical disks that refuse access to all usable sectors due to a
-single bad sector. Such a disk would be impractical in many LVM (Logical Volume
-Manager) or RAID configurations. Similarly, the ability for developers who wish
-to use SlateDB to achieve Logical Durability is severely compromised if access
-to uncorrupted data blocks is not allowed.
-
-To support those operators who value durability, the developer making the call
-to `get` can decide to abort the application if and corruption error is returned.
-The application can then provide operators with detailed information included
-in the error returned by SlateDB on the nature of the corruption and suggestions
-on how to fix it.
-
-To support those operators who value availability, the error can be logged or
-ignored until either the database is deleted, or corruption repaired.
-
-To support repair of the SST, I propose we introduce new methods called
-`DB::sst_check()` , `DB::sst_inspect()` and `DB::sst_repair()` providing
-developers tools to validate the integrity of the entire database, inspect the
-current contents of a specific SST and repair a specific SST which has been
-identified as corrupt.
-
-> NOTE: We could optionally support verification of the entire database when
-> `DB::open_with_opts()` is called via some future `DBOptions` config. In
-> addition, we can expose the inspection and repair methods via a CLI available
-> to operators.
-
-
+in the context of durable object storage, it is not impossible.
+
+To ensure durability by default, I propose that `open_with_opts()` perform
+WAL recovery synchronously and return any WAL corruption found 
+as errors, aborting the database open process until the corruption is resolved.
+The error returned should be a distinct type that can easily be matched to 
+determine if the error is due to non-transient corruption or a transient error
+(such as network connectivity issues).
+
+#### Handling SST Corruption During Compaction
+Compaction is an asynchronous process that may encounter corrupted SST files
+during operation. If corruption is detected, the compaction process defers
+control to the `DBOptions.on_corruption` callback.
+
+- If the `DBOptions.on_corruption` callback returns an error, the compaction
+  operation aborts the current run.
```

Comment:
Yes.. however, my thought here (especially in light of your previous comment `Return errors or call on_corruption() callback?`) is to maintain semantic simplicity for `DBOptions.on_corruption`. If `on_corruption` returns an error for any operation then we should abort the operation.

We might be able to make an exception for this scenario. 🤔

Reply:
Thinking about this more, My gut is to maintain semantic simplicity as I can imagine we might discover several different scenarios where we could (should?) ignore the returned error from `DBOptions.on_corruption` and "do the right thing". I have no strong opinions on this either way.

Reply:
I don't have a strong opinion either. In this scenario, we could split hairs about the semantics of on_corruption: is it on_corruption while compacting a set of SRs or is it on_corruption while running an entire compaction task?

🤔 Given that compaction plans and strategies are pluggable, now that I think it about it, it might actually _not_ be safe in all scenarios to skip the bad SRs. I recall @agavra discussing some of Scylla's strategies that wouldn't work for this.

In light of this, I think you're right--let's keep it as-is. If a compaction detects a corrupt file, it aborts. This will cause data to back up indefinitely on object storage, and could lead to manifest file bloat and slower reads. This seems preferable to potentially breaking things.